### PR TITLE
Move learning rate and releated op to pserver

### LINKED ISSUE #8032: Adapt learning rate decay with Fluid distributed training
2*trainers + 2*pservers, following is pserver logs:

```bash
E0201 06:20:34.945315  5910 listen_and_serv_op.cc:156] run sub program error enforce framework::product(lr_dims) == 1 failed, 0 != 1
Learning rate should have 1 element at [/paddle/paddle/operators/sgd_op.cc:36]
PaddlePaddle Call Stacks:
0       0x7f3a9cecfde7p paddle::platform::EnforceNotMet::EnforceNotMet(std::__exception_ptr::exception_ptr, char const*, int) + 727
1       0x7f3a9d4d4eaap paddle::operators::SGDOp::InferShape(paddle::framework::InferShapeContext*) const + 986
2       0x7f3a9d57602fp paddle::framework::OperatorWithKernel::Run(paddle::framework::Scope const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) const + 2527
3       0x7f3a9cf95630p paddle::framework::Executor::Run(paddle::framework::ProgramDesc const&, paddle::framework::Scope*, int, bool, bool) + 1504
4       0x7f3a9d51e143p paddle::operators::ListenAndServOp::Run(paddle::framework::Scope const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) const + 3443
5       0x7f3a9cf95630p paddle::framework::Executor::Run(paddle::framework::ProgramDesc const&, paddle::framework::Scope*, int, bool, bool) + 1504
6       0x7f3a9ceeeac3p void pybind11::cpp_function::initialize<pybind11::cpp_function::initialize<void, paddle::framework::Executor, paddle::framework::ProgramDesc const&, paddle::framework::Scope*, int, bool, bool, pybind11::name, pybind11::is_method, pybind11::sibling>(void (paddle::framework::Executor::*)(paddle::framework::ProgramDesc const&, paddle::framework::Scope*, int, bool, bool), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&)::{lambda(paddle::framework::Executor*, paddle::framework::ProgramDesc const&, paddle::framework::Scope*, int, bool, bool)#1}, void, paddle::framework::Executor*, paddle::framework::ProgramDesc const&, paddle::framework::Scope*, int, bool, bool, pybind11::name, pybind11::is_method, pybind11::sibling>(pybind11::cpp_function::initialize<void, paddle::framework::Executor, paddle::framework::ProgramDesc const&, paddle::framework::Scope*, int, bool, bool, pybind11::name, pybind11::is_method, pybind11::sibling>(void (paddle::framework::Executor::*)(paddle::framework::ProgramDesc const&, paddle::framework::Scope*, int, bool, bool), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&)::{lambda(paddle::framework::Executor*, paddle::framework::ProgramDesc const&, paddle::framework::Scope*, int, bool, bool)#1}&&, void (*)(paddle::framework::Executor*, paddle::framework::ProgramDesc const&, paddle::framework::Scope*, int, bool, bool), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&)::{lambda(pybind11::detail::function_call&)#3}::_FUN(pybind11::detail::function_call) + 579
7       0x7f3a9ceec754p pybind11::cpp_function::dispatcher(_object*, _object*, _object*) + 1236
8             0x4cad00p PyEval_EvalFrameEx + 28048
9             0x4c2705p PyEval_EvalCodeEx + 597
10            0x4ca088p PyEval_EvalFrameEx + 24856
11            0x4c9d7fp PyEval_EvalFrameEx + 24079
12            0x4c2705p PyEval_EvalCodeEx + 597
13            0x4c24a9p PyEval_EvalCode + 25
14            0x4f19efp
15            0x4ec372p PyRun_FileExFlags + 130
16            0x4eaaf1p PyRun_SimpleFileExFlags + 401
17            0x49e208p Py_Main + 1736
18      0x7f3ae5dd8830p __libc_start_main + 240
19            0x49da59p _start + 41
```

### PR DESCRIPTION
**Pull request by Yancey1989**
Fixed #8032

### CODE REVIEW

File: python/paddle/v2/fluid/distribute_transpiler.py [hunk 1]
```diff
@@ -421,23 +464,89 @@ def _append_pserver_ops(self, program, pserver_program, opt_op, endpoint):
...
+            for _, vars in opt_op.inputs.iteritems():
+                if not isinstance(vars, list):
+                    vars = [vars]
+                for var in vars:
+                    program.global_block().create_var(
+                        name=var.name,
+                        persistable=var.persistable,
+                        dtype=var.dtype,
+                        shape=var.shape)
+                    pserver_program.global_block().create_var(
```
Comment (typhoonzero) [inline python/paddle/v2/fluid/distribute_transpiler.py]: Can add a FIXME here, should only create var once.

File: python/paddle/v2/fluid/distribute_transpiler.py [hunk 2]
```diff
@@ -421,23 +464,89 @@ def _append_pserver_ops(self, program, pserver_program, opt_op, endpoint):
...
+            for j in xrange(i, len(optimize_ops)):
+                op1 = optimize_ops[i]
+                op2 = optimize_ops[j]
+                if self._is_op_connected(op1, op2):
+                    ufind.union(op1, op2)
+        return ufind
+
+    def _is_opt_op(self, op):
+        # optimize op: SGDOptimize, MomentumOptimizer, AdamOptimizer and etc... 
+        if op.inputs and op.inputs.has_key("Param") \
```
Comment (typhoonzero) [inline python/paddle/v2/fluid/distribute_transpiler.py]: Add comment HACK, this is a hack.

File: python/paddle/v2/fluid/distribute_transpiler.py [hunk 3]
```diff
@@ -457,27 +566,33 @@ def get_pserver_program(self, endpoint):
...
-            is_op_on_pserver = self._is_op_on_pserver(endpoint,
-                                                      self.optimize_ops, idx)
-            if not is_op_on_pserver:
+        opt_op_on_pserver = []
+        for _, op in enumerate(self.optimize_ops):
+            if self._is_opt_op(op) and self._is_opt_op_on_pserver(endpoint, op):
+                opt_op_on_pserver.append(op)
+
+        for _, op in enumerate(self.optimize_ops):
+            if not True in [
```
Comment (typhoonzero) [inline python/paddle/v2/fluid/distribute_transpiler.py]: `not True in []` is still confusing.

File: python/paddle/v2/fluid/distribute_transpiler.py [hunk 4]
```diff
@@ -457,27 +566,33 @@ def get_pserver_program(self, endpoint):
...
-                print("create variable for program: %s.trainer_%d" %
-                      (v.name, trainer_id))
                 pserver_program.global_block().create_var(
                     name="%s.trainer_%d" % (v.name, trainer_id),
                     persistable=True,
                     dtype=v.dtype,
                     shape=v.shape)
         # step6
         optimize_sub_program = Program()
+        ufind = self._create_ufind(self.optimize_ops)
```
Comment (typhoonzero) [inline python/paddle/v2/fluid/distribute_transpiler.py]: Add comments or document about why we need `ufind`.

File: python/paddle/v2/fluid/distribute_transpiler.py (hunk 1 above)
Comment (Yancey1989) [reply python/paddle/v2/fluid/distribute_transpiler.py]: Done.

File: python/paddle/v2/fluid/distribute_transpiler.py (hunk 2 above)
Comment (Yancey1989) [reply python/paddle/v2/fluid/distribute_transpiler.py]: Done.

File: python/paddle/v2/fluid/distribute_transpiler.py (hunk 4 above)
Comment (Yancey1989) [reply python/paddle/v2/fluid/distribute_transpiler.py]: Done.

File: python/paddle/v2/fluid/distribute_transpiler.py (hunk 3 above)
Comment (Yancey1989) [reply python/paddle/v2/fluid/distribute_transpiler.py]: Done.

**Comment by Yancey1989** [comment]
I'll separate **return all optimize and related ops**  feature into another PR #8280

**Pull request merged**